cli/command/image/tree.go

```diff
@@ -0,0 +1,296 @@
+package image
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/morikuni/aec"
+
+	"github.com/containerd/platforms"
+	"github.com/docker/docker/api/types/filters"
+	imagetypes "github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/go-units"
+)
+
+type treeOptions struct {
+	all     bool
+	filters filters.Args
+}
+
+func runTree(ctx context.Context, dockerCLI command.Cli, opts treeOptions) error {
+	images, err := dockerCLI.Client().ImageList(ctx, imagetypes.ListOptions{
+		All:            opts.all,
+		ContainerCount: true,
+		Filters:        opts.filters,
+		Manifests:      true,
+	})
+	if err != nil {
+		return err
+	}
+
+	view := make([]topImage, 0, len(images))
+	for _, img := range images {
+		details := imageDetails{
+			ID:        img.ID,
+			DiskUsage: units.HumanSizeWithPrecision(float64(img.Size), 3),
+			Used:      img.Containers > 0,
+		}
+
+		children := make([]subImage, 0, len(img.Manifests))
+		for _, im := range img.Manifests {
+			if im.Kind != imagetypes.ManifestKindImage {
+				continue
+			}
+
+			imgData := im.ImageData
+			platform := imgData.Platform
+
+			sub := subImage{
+				Platform:  platforms.Format(platform),
+				Available: im.Available,
+				Details: imageDetails{
+					ID:        im.ID,
+					DiskUsage: units.HumanSizeWithPrecision(float64(im.Size.Total), 3),
+					Used:      len(imgData.Containers) > 0,
+				},
+			}
+
+			children = append(children, sub)
+		}
+
+		view = append(view, topImage{
+			Names:    img.RepoTags,
+			Details:  details,
+			Children: children,
+		})
+	}
+
+	return printImageTree(dockerCLI, view)
+}
+
+type imageDetails struct {
+	ID        string
+	DiskUsage string
+	Used      bool
+}
+
+type topImage struct {
+	Names    []string
+	Details  imageDetails
+	Children []subImage
+}
+
+type subImage struct {
+	Platform  string
+	Available bool
+	Details   imageDetails
+}
+
+const headerSpacing = 3
+
+//nolint:gocyclo
+func printImageTree(dockerCLI command.Cli, images []topImage) error {
+	out := dockerCLI.Out()
+	_, width := out.GetTtySize()
+	if width == 0 {
+		width = 80
+	}
+	if width < 20 {
+		width = 20
+	}
+
+	headers := []header{
+		{Title: "Image", Width: 0, Left: true},
+		{Title: "ID", Width: 12},
+		{Title: "Disk usage", Width: 10},
+		{Title: "Used", Width: 4},
+	}
+
+	nameWidth := int(width)
+	for idx, h := range headers {
+		if h.Width == 0 {
+			continue
+		}
+		d := h.Width + headerSpacing
+		// If the first column gets too short, remove remaining columns
+		if nameWidth-d < 12 {
+			headers = headers[:idx]
+			break
+		}
+		nameWidth -= d
+	}
+
+	maxImageName := len(headers[0].Title)
+	for _, img := range images {
+		for _, name := range img.Names {
+			if len(name) > maxImageName {
+				maxImageName = len(name)
+			}
+		}
+		for _, sub := range img.Children {
+			if len(sub.Platform) > maxImageName {
+				maxImageName = len(sub.Platform)
+			}
+		}
+	}
```

Comment:
These bits of the code are a bit confusing to me – I'm okay merging as it's experimental, but I wonder if we can subtract some of these bits (like call this `calcMaxImageNameLength`, and others) and add some tests for them. OTOH, if we want to merge just to get them out and I'm fine as long as we refactor/merge before making it no longer experimental.

Reply:
Eventually we should replace all this custom formatting code with something like lipgloss, but I didn't want to introduce a new dependency just yet.
We can probably do that once we start implementing Javier's designs for other commands (and decide which library we want to settle with).

Reply:
I think that's fine eventually, but until then I'd prefer to have a nicer implementation – I don't think we should wait on that to at least do a small refactor.

Reply:
Something like this: https://github.com/laurazard/cli/blob/b20c2598ca71e3c0db6247721869a5476d907fff/cli/command/image/tree.go#L97-L252

Reply:
(this also removes the need for that `//nolint:gocyclo` directive).

Reply:
Yeah, not saying to not refactor it now 😅 Will do it!